### PR TITLE
Cleanup after GC info version update.

### PR DESCRIPTION
**Pull request by VSadov**
There should not be any functional changes. The most affected part is GC stress as there is a lot of dead code there now.
Some code moved around to tease apart reachable and unreachable scenarios.

- [x] remove code specific to no longer used GC info versions
- [x] separate major parts that deal with X86 specific scenarios from everything else
- [x] once we have X86 and "everything else", we can make "everything else" much simpler. 
I.E. the figuring the targets of calls and what they may return is no longer needed on non-X86 - neither at instrumentation time nor at fault time.
- [x] once separated, X86 part can be somewhat simpler too. (no multireg returns, etc)
- [x] look for more things, address TODOs

### CODE REVIEW

**Comment by VSadov** [comment]
I think this is ready for review.
Cc: @jkotas @janvorli @jakobbotsch

File: src/coreclr/vm/gccover.cpp [hunk 1]
```diff
@@ -162,6 +95,9 @@ void SetupAndSprinkleBreakpoints(
 {
     _ASSERTE(!nativeCodeVersion.IsNull());
 
+    // CONSIDER: does anyone call this with fZapped == true ? are there plans?
```
Comment (jkotas) [inline src/coreclr/vm/gccover.cpp:98]: `Zap` was a code name for fragile NGen.

Does the GC stress instrumentation work for R2R code? If it does, this flag can be deleted. If it does not, what needs to happen to enable the GC stress instrumentation for R2R code?

Comment (VSadov) [reply src/coreclr/vm/gccover.cpp:98]: I have two theories - either R2R code comes on the JIT path (and JIT does not really JIT much, maybe some fixups), then we could remove `fZapped ` path as a part this change.
Or we do not stress R2R. There is nothing that would make that impossible, I think R2R code is writeable in the same sense as JIT code is writeable. Enabling stress over R2R would be a separate change though. 

I will check what happens with R2R.

**Comment by janvorli** [comment]
@VSadov I wonder if it would be possible to reorder stuff in the gccover.cpp so that the diff github shows would make sense. Looking at it now, it is very difficult to see what was changed and what was just moved.

**Comment by VSadov** [comment]
> @VSadov I wonder if it would be possible to reorder stuff in the gccover.cpp so that the diff github shows would make sense. Looking at it now, it is very difficult to see what was changed and what was just moved.

Some of that is because a several functions make sense only for x86, and a few are very different between x86 and "the rest", so I brought x86-specific parts closer to each other under one `#if defined(TARGET_X86)`. I think it is easier to work with this code by concentrating on X86 or "everything else" separately. Moving code did contribute to extra diffs. 
I'll see if I could rearrange the X86 part a bit so diffs are smaller.

File: src/coreclr/vm/gccover.cpp [hunk 2]
```diff
@@ -1845,29 +1207,172 @@ void DoGcStress (PCONTEXT regs, NativeCodeVersion nativeCodeVersion)
...
+// A managed thread (T) can race with the GC as follows:
+// 1) Current thread T is in preemptive mode, GC happens, starts scanning T's stack.
+// 2) The Jitted code puts thread T to cooperative mode, as part of PInvoke epilog.
+// 3) Jitted code checks g_TrapReturningThreads, calls CORINFO_HELP_STOP_FOR_GC()
+// 4) If any of the code that does #3 is interruptible, we may hit GCStress trap and start
+//    another round of GCStress while in Cooperative mode.
+// 5) Now, thread T can modify the stack (ex: RedirectionFrame setup) while the GC thread is scanning it.
+//
+// This race is now mitigated below. Where we won't initiate a stress mode GC
+// for a thread in cooperative mode with an active ICF, if g_TrapReturningThreads is true.
```
Comment (VSadov) [inline src/coreclr/vm/gccover.cpp:1305]: The original explanation for the g_TrapReturningThreads + active ICF scenario causing a race was somewhat questionable, so I initially removed the mitigation, but the race appears to be real. 

Corrected the explanation of how the race may happen and brought the mitigation back.

**Comment by VSadov** [comment]
> Does the GC stress instrumentation work for R2R code? If it does, this flag can be deleted. If it does not, what needs to happen to enable the GC stress instrumentation for R2R code?

It does not look like we instrument R2R code for GC stress. The instrumentation (`SetupGcCoverage`) is called from `JitCompileCodeLocked` and that happens only on JIT-compiled code path.

I will log a follow-up issue. We should be able to GC-stress R2R code.

**Comment by VSadov** [comment]
> I'll see if I could rearrange the X86 part a bit so diffs are smaller.

@janvorli - We had a lot of platform specific code and now a lot of that is gone, while X86 specific code is still in. Moving code roughly to where it was resulted in many `#if defined(TARGET_X86)`, with no corresponding "everything else" code. 

I tried to move code around to reduce diffs, but it made things mostly worse. The diffs were still big enough that VS differ could not make a lot of sense of it (I assume github diff will not be any better). So it basically the changes were making the file harder to read without much improvement to the diff.

The code that I moved generally stayed the same, modulo removing dead code. I was particularly careful that X86 code path did not change. In couple places I made separate X86 and non-X86 copies (`DoGcStress`, `GCCoverageInfo::SprinkleBreakpoints`) - just to make sure that X86 stays the same. 

I think the easiest way to review this change is basically by reading the new code, and when in doubt finding old version and check with that. 

I think the most "interesting" part is that AMD64 code path is now merged with ARM,RISC,etc.. 
That is where I had to add some code - like to figure lengths of x64 instructions we need to ask disassembler. However we no longer go over all the instructions for the function/funclets, we just iterate over safe points or over interruptible ranges - the same as we did on RISC architectures.

The part that the stress runs are green and take roughly the same time as before give me a good deal of confidence that the code is functionally the same as before.
(and I did verify in debugger that we actually stress things :-)

**Comment by VSadov** [comment]
Rebased onto recent main.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by VSadov** [comment]
Thanks!!

**Pull request merged**